Make mpConfig-3.0 compatible with EE10

Azquelt opened the pull request:
This PR includes some prerequisite work:
* Creating the `mpCompatible-6.0` feature
* Adding an MP60 feature set to MicroProfileActions
* Re-factoring the EE compatibility tests to make it easier to express that we don't yet want to assert the compatibility of the new MP60 features against EE9 or EE10

For #21968

tevans78 committed:
Create the mpCompatible-6.0 feature

Currently this depends on eeCompatible-10.0 and tolerates 9.0, but
before release it will be updated to not tolerate 9.0.

This needs to be beta because we have MP 6.0 features which are already
in beta.

tevans78 committed:
Make mpConfig-3.0 tolerate EE10 and MP6

mpConfig-3.0 will be included in both MP5 and MP6

Azquelt committed:
Run MP Config 3.0 tests with EE10 features

Update test repeats using MP50 to MP60 (both contain mpConfig-3.0)

tevans78 committed:
Add MP 6 features to MicroProfileActions

Azquelt committed:
Refactor EE compatibility tests

Refactor to make the test setup easier to understand and modify as we
bring more features into compatibility with EE10.

Azquelt commented:
#build
#libby

Azquelt commented:
Not planning to merge this until after gm-22.0.0.9 is tagged

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/BasicConfigTests.java:
```diff
@@ -76,9 +77,14 @@ public class BasicConfigTests extends FATServletClient {
 
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP12, MicroProfileActions.MP13, MicroProfileActions.MP14,
+    @ClassRule
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60,
```
MicroProfileActions.repeat will need an update to work on Java 8 because it marks the first FeatureSet for lite and the rest for full.  But with Java 8 then no tests will run and the build will break.

Otherwise you do what we have done in other FATs if Java 8 use MP 5 as first feature set and don't have mp5 in the other feature sets lists.  If I am not making sense, let me know.

Azquelt replied:
Ah, good point, thank you

Azquelt replied:
Updated, I've added a new `MicroProfileActions.repeat` method which substitutes the most recent compatible set from `otherFeatureSets`.

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/ClassLoadersTest.java:
```diff
@@ -47,7 +47,7 @@ public class ClassLoadersTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP14, MicroProfileActions.MP41);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP14, MicroProfileActions.MP41);
```
Should MP50 be added to the end of the method?

Azquelt replied:
I was not going to because the MP config implementation is the same for MP50 and MP60, so it doesn't seem necessary to run both, even in full mode, for functionality which is wholly within MP config, though I was going to run that past Tom once he's back.

The tests I was going to keep running on both were:
* the `BasicConfTests` test above
* the `mpConfig-3.0` TCK (once that's fixed to work against EE10)

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/DefaultSourcesTest.java:
```diff
@@ -52,7 +52,7 @@ public class DefaultSourcesTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP33, MicroProfileActions.MP41);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP33, MicroProfileActions.MP41);
```
Should MP50 be added to the end of the method?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/SharedLibTest.java:
```diff
@@ -50,7 +50,7 @@ public class SharedLibTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP20, MicroProfileActions.MP41);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP20, MicroProfileActions.MP41);
```
Should MP50 be added to the end of the method?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/SimultaneousRequestsTest.java:
```diff
@@ -52,7 +52,7 @@ public class SimultaneousRequestsTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP33, MicroProfileActions.MP41);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP33, MicroProfileActions.MP41);
```
Should MP50 be added to the end of the method?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/OrdinalsForDefaultsTest.java:
```diff
@@ -50,7 +50,7 @@ public class OrdinalsForDefaultsTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP20, MicroProfileActions.MP41);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP20, MicroProfileActions.MP41);
```
Should MP50 be added to the end of the method?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/StressTest.java:
```diff
@@ -49,7 +49,7 @@ public class StressTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP14, MicroProfileActions.MP41);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP14, MicroProfileActions.MP41);
```
Should MP50 be added to the end of the method?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.2_fat/fat/src/com/ibm/ws/microprofile/config12/test/Config12ConverterTests.java:
```diff
@@ -50,7 +50,7 @@ public class Config12ConverterTests extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP41, MicroProfileActions.MP13);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP41, MicroProfileActions.MP13);
```
Should MP50 be added after MP60?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/ConfigOrdinalServerXMLTest.java:
```diff
@@ -73,7 +73,7 @@ public class ConfigOrdinalServerXMLTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP41, MicroProfileActions.MP20);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP41, MicroProfileActions.MP20);
```
Should MP50 be added after MP60?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/ServerXMLTest.java:
```diff
@@ -65,7 +65,7 @@ public class ServerXMLTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP41, MicroProfileActions.MP14);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP41, MicroProfileActions.MP14);
```
Should MP50 be added after MP60?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java:
```diff
@@ -65,7 +65,7 @@ public class VariableServerXMLTest extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP41, MicroProfileActions.MP20);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP41, MicroProfileActions.MP20);
```
Should MP50 be added after MP60?

jhanders34 commented on dev/com.ibm.ws.microprofile.config.1.4_fat/fat/src/com/ibm/ws/microprofile/config14/test/Config14Tests.java:
```diff
@@ -40,7 +40,7 @@ public class Config14Tests extends FATServletClient {
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP50, MicroProfileActions.MP33, MicroProfileActions.LATEST);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP33, MicroProfileActions.LATEST);
```
LATEST has changed meaning.  Do we need MP41 and MP50 both covered?

Azquelt replied:
I'm confused by LATEST, I'm not sure why it's still MP41 when MP50 has been out for a while and when I looked at existing tests, many are running against LATEST and MP50, so they'd all need updated as well.

For that reason I haven't changed LATEST in this PR, and unless we can work out a way for this situation to not happen again, I'd advocate for removing it altogether.

For this particular class, I didn't think it was necessary to run against both MP60 and MP50.

Azquelt committed:
Ensure we always run at least one repeat

Often we just run one repeat of a test in LITE mode. If that featureSet
isn't compatible with the current java version, we will skip it.

In this situation we should instead try to find another candidate to be
the first feature set.

Azquelt commented:
#build #libby

jhanders34 commented on dev/fattest.simplicity/src/componenttest/rules/repeater/MicroProfileActions.java:
```diff
@@ -368,7 +388,46 @@ public static RepeatTests repeat(String server, FeatureSet firstFeatureSet, Feat
+
+            if (newestSupportedSet.isPresent()) {
+                firstFeatureSet = newestSupportedSet.get();
+                otherFeatureSets = new HashSet<>(otherFeatureSets);
```
This will work until we have a MP 6.1 or 7.0 right?  Don't we need to remove the ones that will not work with the current JDK from the otherFeatureSets?

jhanders34 replied:
It could be done in the RepeatActions.repeat method that if the MP level is not supported on the current JDK in the other featureSets, we just do do the `andWith` call for that MP level.

Azquelt replied:
I thought the problem here was just that if a feature set wasn't supported on the current JDK, it would be skipped because `withMinJavaLevel()` was set on the repeat action.

The only problem then was when the first feature set gets skipped because that means we don't run any tests at all in lite mode.

jhanders34 replied:
Oh yeah.  It was late.  :)

jhanders34 commented on dev/io.openliberty.jakartaee10.internal_fat/fat/src/io/openliberty/jakartaee10/internal/tests/EE10FeatureCompatibilityTest.java:
```diff
@@ -92,114 +90,80 @@ public class EE10FeatureCompatibilityTest extends FATServletClient {
+        // MP features are only compatible if they're in MP versions which work with EE10
+        compatibleFeatures.removeAll(FeatureUtilities.allMpFeatures());
+        // Add only the MP 6.0 features which we've made compatible
+        compatibleFeatures.add("mpConfig-3.0");
```
I believe with you adding mpConfig-3.0 being compatible mpRestClient-3.0 should also work.

Azquelt replied:
That's possibly true, I hadn't added it here because that wasn't something I was trying to make work, but as you've pointed it out I can add it now.

Azquelt replied:
With the current changes, mpRestClient-3.0 is not yet compatible:
```
Got unexpected conflict for mpRestClient-3.0 [io.openliberty.servlet.api, com.ibm.websphere.appserver.eeCompatible, io.openliberty.jsonp]
```

I _think_ it may just need tolerations added to the API feature which were in #21903 and which I sent to Jim yesterday: https://github.com/OpenLiberty/open-liberty/commit/d9c14888052102be7f70e2bd4de761bcabd0c66d

jhanders34 replied:
oh ok.  That is easy enough.

jhanders34 commented on dev/io.openliberty.jakartaee10.internal_fat/fat/src/io/openliberty/jakartaee10/internal/tests/FeatureUtilities.java:
```diff
@@ -0,0 +1,166 @@
+        testFeatures.add("txtest-2.0");
+        testFeatures.add("ejbTest-1.0");
+        testFeatures.add("ejbTest-2.0");
+        testFeatures.add("enterpriseBeansTest-2.0");
```
Not sure why you added enterpriseBeansTest-2.0.  It is marked IBM-Test-Feature, so it wouldn't be in the list already.

jhanders34 replied:
oh I see why

jhanders34 commented on dev/io.openliberty.jakartaee10.internal_fat/fat/src/io/openliberty/jakartaee10/internal/tests/EE10FeatureCompatibilityTest.java:
```diff
@@ -436,7 +388,11 @@ private void checkFeatures(String baseFeature, Queue<String> featureQueue, Map<S
             Log.info(c, "checkFeatures", "start testing: " + feature);
             featuresToTest.set(1, feature);
-            boolean expectToConflict = nonEE10JavaEEFeatures.contains(feature) || nonEE10MicroProfileFeatures.contains(feature) || incompatibleValueAddFeatures.contains(feature);
+            if (!compatibleFeatures.contains(feature) && !incompatibleFeatures.contains(feature)) {
+                // Don't test this feature
+                continue;
+            }
```
Won't this cause us to possibly miss features.  There was a reason I didn't have this logic.

Azquelt replied:
Yes, and I can see why it was written like that but I want to change it for two reasons:

1. During development when we're trying to make features compatible with EE10, it's useful to be able to say "I don't want to assert compatibility of this feature either way yet. If I can't do this then as I change my feature to make it compatible with EE10, I not only need to get my own FAT tests repeating and passing, I also need to change this test to expect the change I'm making (and any other knock-on changes, maybe making my feature compatible has also made another feature compatible that I wasn't aware of). In practise, that means I need to run this test, see what fails and change the test to expect the actual result.

2. With the way that `getIncompatibleFeatures()` is written, it's very easy to see which features aren't included in either set. The starting point is that everything not compatible is incompatible, and then we take some features out that we don't want to assert incompatibility for. As we finish development on MP6, it should be easy to verify that we're not missing any features.

jhanders34 commented on dev/io.openliberty.jakartaee10.internal_fat/fat/src/io/openliberty/jakartaee10/internal/tests/EE10FeatureCompatibilityTest.java:
```diff
@@ -474,8 +430,8 @@ public void transportSecurityUsed() {
-        for (String feature : features) {
-            if (nonEE10JavaEEFeatures.contains(feature) || nonEE10MicroProfileFeatures.contains(feature) || incompatibleValueAddFeatures.contains(feature)) {
+        for (String feature : allFeatures) {
+            if (!compatibleFeatures.contains(feature)) {
```
Is there a reason you switched this from incompatibleFeatures.contains to !compatibleFeatures.contains?

Azquelt replied:
It's logically correct for what the test was doing (assuming I've understood it correctly).

`compatibleFeatures` are those that we want to assert are compatible
`incompatibleFeatures` are those that we want to assert are incompatible

This test is checking that features are compatible (i.e. that they don't cause conflicts), therefore it should only test features which are in the `compatibleFeatures` set.

jhanders34 commented on dev/io.openliberty.jakartaee10.internal_fat/fat/src/io/openliberty/jakartaee10/internal/tests/EE10FeatureCompatibilityTest.java:
```diff
@@ -474,8 +430,8 @@ public void transportSecurityUsed() {
         ee10FeaturesThatEnableSsl.add("connectorsInboundSecurity-2.0");
```
This is not an EE10 feature any longer so it can be removed.

Azquelt replied:
Ok, I will take that out while I'm here.

Azquelt replied:
Done

jhanders34 commented on dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java:
```diff
@@ -92,114 +91,74 @@ public class EE9FeatureCompatibilityTest extends FATServletClient {
+        incompatibleFeatures.removeAll(compatibleFeatures);
 
-        return compatFeatures;
+        // ...but there are a few in-development features which are maybe compatible
+        incompatibleFeatures.removeAll(MicroProfileActions.MP60.getFeatures());
```
Is this really needed?

Azquelt replied:
It is at the moment. `mpOpenAPI-3.1` and `mpTelemetry-1.0` are currently compatible with EE9, though they will not be by the time they're ready to release.

jhanders34 commented on dev/io.openliberty.jakartaee10.internal_fat/fat/src/io/openliberty/jakartaee10/internal/tests/EE10FeatureCompatibilityTest.java:
```diff
@@ -474,8 +430,8 @@ public void transportSecurityUsed() {
 
-        for (String feature : features) {
-            if (nonEE10JavaEEFeatures.contains(feature) || nonEE10MicroProfileFeatures.contains(feature) || incompatibleValueAddFeatures.contains(feature)) {
+        for (String feature : allFeatures) {
+            if (!compatibleFeatures.contains(feature)) {
                 continue;
             }
```
```suggestion
        for (String feature : compatibleFeatures) {
```

Didn't have a compatibility list previously so had to iterate over all features and ignore incompatible features.  But with having a compatibility list can just iterate over those.

Azquelt replied:
Ok, that does make way more sense

jhanders34 replied:
The change needs to be applied to EE9 test as well.  I didn't state it earlier, but any comments I made in the EE10 test apply to the EE9 test as well.

Azquelt replied:
Done

Azquelt committed:
Simplify loop in transportSecurityUsed

Just loop through the compatible features, rather than looping through
all features and dropping the ones that aren't compatible.

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
Remove old reference to non-EE10 feature

connectorsInboundSecurity-2.0 is not an EE10 feature any longer

Azquelt commented:
#libby
#build

Azquelt commented:
Missed that the previous personal build failed with a compile error in `MPContextPropActions`, now fixed

#build
#libby

Azquelt commented:
Failures in the last build _look_ unrelated to this change but don't match existing defects, running again to be sure.
#build

Azquelt commented:
I've realised that the change in `MicroProfileActions` may change the order in which repeats are done and that this may break some tests when run in FULL mode. For this reason I'm going to back out this change and redeliver it with a fix.
